db/blob/blob_source.cc

```diff
@@ -0,0 +1,295 @@
+      Slice value(static_cast<char*>(blob_cache->Value(cache_handle)));
+      blob->SetCachedValue(&value, blob_cache, cache_handle);
+      return Status::OK();
+    }
+  }
+
+  assert(blob->IsEmpty());
+
+  // TODO: If not found, search from the compressed blob cache.
+
+  return Status::NotFound("Blob record not found in cache");
+}
+
+Status BlobSource::PutBlobToCache(const Slice& cache_key, Cache* blob_cache,
+                                  CachableEntry<Slice>* cached_blob,
+                                  PinnableSlice* blob) const {
+  assert(blob);
+  assert(!cache_key.empty());
+
+  const Cache::Priority priority = Cache::Priority::LOW;
+
+  Status s;
+
+  // TODO: Insert compressed blob into compressed blob cache.
+  // Release the hold on the compressed cache entry immediately.
+
+  // Insert into uncompressed blob cache
+  if (blob_cache != nullptr) {
+    Cache::Handle* cache_handle = nullptr;
+    s = InsertEntryToCache(lowest_used_cache_tier_, blob_cache, cache_key,
+                           GetCacheItemHelper(), blob, blob->size(),
+                           &cache_handle, priority);
+    if (s.ok()) {
+      assert(cache_handle != nullptr);
+      cached_blob->SetCachedValue(blob, blob_cache, cache_handle);
+    }
+  }
+
+  return s;
+}
+
+CacheKey BlobSource::GetCacheKey(uint64_t file_number, uint64_t file_size,
+                                 uint64_t offset) const {
+  OffsetableCacheKey base_cache_key =
+      OffsetableCacheKey(db_id_, db_session_id_, file_number, file_size);
+  return base_cache_key.WithOffset(offset);
+}
+
+Status BlobSource::GetBlob(const ReadOptions& read_options,
+                           const Slice& user_key, uint64_t file_number,
+                           uint64_t offset, uint64_t file_size,
+                           uint64_t value_size,
+                           CompressionType compression_type,
+                           FilePrefetchBuffer* prefetch_buffer,
+                           PinnableSlice* value, uint64_t* bytes_read) {
+  assert(value);
+
+  const uint64_t key_size = user_key.size();
+
+  if (!IsValidBlobOffset(offset, key_size, value_size, file_size)) {
+    return Status::Corruption("Invalid blob offset");
+  }
+
+  const CacheKey cache_key = GetCacheKey(file_number, file_size, offset);
+
+  CachableEntry<Slice> blob_entry;
+
+  // TODO: We haven't support cache tiering for blob files yet, but will do it
+  // later.
+  if (blob_cache_ &&
+      lowest_used_cache_tier_ != CacheTier::kNonVolatileBlockTier) {
+    const Status s =
+        MaybeReadBlobFromCache(prefetch_buffer, read_options, cache_key, offset,
+                               true /* wait_for_cache */, &blob_entry);
+
+    if (s.ok() && blob_entry.GetValue() != nullptr) {
+      assert(blob_entry.GetValue()->size() == value_size);
+      if (bytes_read) {
+        *bytes_read = value_size;
+      }
+      value->PinSelf(*blob_entry.GetValue());
+      blob_entry.TransferTo(value);
```

Comment:
I have a feeling that we need only one of the two (either copy the value, or transfer ownership of the cache handle).

Reply:
Sure, we can copy first since there is an another task for that.